### DISCUSSION TITLE
how to start the credentials file?

### DISCUSSION THREAD

**Discussion by alwaysftd** [areed1192/td-ameritrade-python-api#175] [Q&A]
Hi all:

I'm just getting started and I'm having trouble with the credentials file.  I'm specifying a path but no file gets created at the path I specify.  Am I supposed to create the JSON file myself first?  How does this work?  I've tried single backslashes, double backslashes, and forward slashes and nothing seems to create the file.  I do get the challenge/response from TD Amitrade but nothing involving the credentials gets saved and my next line to request quotes fails with a credentials path error.

```python
def TestConnect():

    TDSession = TDClient(
        client_id='ABCDEF',
        redirect_uri='http://localhost',
        credentials_path=r'c:/test/td_state.json'
    )

    TDSession.login()
    msft_quotes = TDSession.get_quotes(instruments=['MSFT'])
```

**Comment by areed1192** [comment]
Quick question, do you actually have a folder in your C Drive that is called `test`? If not that's probably the reason.

**Comment by alwaysftd** [comment]
I do, I also tried creating a blank .json file in that location as well.

**Reply by areed1192** [reply]
Yeah you need to delete the blank JSON file or else it will try to load that and it will fail. I would recommend just have the folder there and then make sure there is no JSON file inside of it.

**Reply by Captain-Science** [reply]
I have the same problem.  Here's what I've learned so far.
The path **in the app** must include a filename or I get a permissions error--probably because it thinks the directory is a file.
However, there doesn't have to be a file in the directory.

I went to the TDA site and got the tokens and put them in the json file, but it gets replaced with a file with nulls (see below) and I get the same error:

{
    "access_token": null,
    "refresh_token": null,
    "logged_in": false
}

If there is no file there at all, it will create a file with the same file with null fields [this is set at 104 in "default values" in client.py]. 

I've used Pycharm and VS Code but I still can't get past this.

**Reply by dougbrowning** [reply]
I still have never got it to work in anything but Thonney.  Here are my notes and script.  Hope it helps.   

https://pypi.org/project/td-ameritrade-python-api/#api-key-and-credentials

rename or move the td_state.json file
run the Login Script 1.py - MUST use Thonny for this script.  PyScripter will not work!
it should give you a URL 
open URL and login to TD
copy the result URL (the whole thing even the local host part) and paste into script
should make a new json file

use this every 90 days to get a new auth code
 rename or delete td_state.json first
 then run below
 copy entire URL (including the localhost part) that you get back after logging in
 it will create a new td_state.json  can use this in PyScripter then


from td.client import TDClient


TDSession = TDClient(
    client_id='00000000000',
    redirect_uri='http://localhost',
    credentials_path= r'C:\Users\username\Documents\Stocks\Python\td_state.json',
    auth_flow='flask'
)


TDSession.login()


msft_quotes = TDSession.get_quotes(instruments=['MSFT'])
print (msft_quotes)

Looks like this in Thonny

>>> %Run 'Login script 1.py'
Please go to URL provided authorize your account: https://auth.tdameritrade.com/auth?response_type=code&redirect_uri=http%3A%2F%2Flocalhost&client_id=0000000000000000000%40AMER.OAUTHAP

Paste the full URL redirect here: https://localhost/?code=shotenedforpost

{'MSFT': {'blahblah}
>>>

**Comment by alwaysftd** [comment]
OK I don't know what I'm doing wrong here.

def TestConnect():
    TDSession = TDClient(
        client_id='ABCDEC',
        redirect_uri='http://localhost',
        credentials_path='c:\\td_state.json'
        )
    TDSession.login()
    msft_quotes = TDSession.get_quotes(instruments=['MSFT'])
    print(msft_quotes)

1.  There is no .json file in my root C: drive.
2.  I get prompted with a URL to visit when I run the script.
3.  I copy the URL out of the command prompt into the browser
4.  It prompts me to login and allow the application to access my account.
5.  It redirects to my 'localhost' which times out and then I copy the URL
and paste it in the console.
6.  I get an error after entering the redirect.

"{'credential_path': 'c:\\td_state.json',
 'message': 'The credential file does not contain expiration times for your
'
            'tokens, please go through the oAuth process.'}"

There is no exception raised.  There is no file located in my root c:
drive.  What am I doing wrong?

**Reply by Vyeche** [reply]
Yeah, I get the same error. I ended up trying a different wrapper. I'm also curious why it doesn't work here on a windows machine.

**Comment by duncanmcdonald** [comment]
I just got this running last night without too much trouble on Windows 10. I had to change the backslashes to forward slashes for windows. Also I find that running python scripts in Thonny gives me better debug messages than the command line does. Hope this helps.

**Comment by alwaysftd** [comment]
OK per @duncanmcdonald I tried forward slashes and again I'm getting the same error.  After a few rounds of that, now I get the error:

"A third-party application may be attempting to make unauthorized access to your account. For help with your account, contact us. For information on securing your account, visit the TD Ameritrade Security page."

When I paste the URL for the setup into the browser.  What happened?  I literally cannot figure out why this will not create a .json file but it never happens.

I have tried the path to the root c:\ drive with:

- backslash
- double backslash
- forward slash
- double forward slash

None of these create a .json file on my c: drive.

**Reply by Cale146** [reply]
@alwaysftd, I've received the "third-party" error before when retrying oAuth several times in a row.  It's TD account specific and from what I've noticed has a wait time involved before it clears up.  One thing that has helped me is deleting and then re-creating the application in the My Apps API area of the developer site.

Regarding the originating problem in this issue, I'm getting the same error as you, which seems to be coming from the validate_token function in the client.py file:
 'message': 'The credential file does not contain expiration times for your tokens, please go through the oAuth process.

I'm using Ubuntu 20 and have tried several things as I thought it might be permission specific:
1 - Setting the path in my user's home directory.
2 - Setting the path to a root directory.
3 - Running my test python file in both regular user and (reluctantly) as root.

None of these steps worked, so I'm not sure what's happening at this point.  If I get some time I may grab the source and see what exactly it's choking on in this function.

**Comment by duncanmcdonald** [comment]
I'm thinking that you have a permissions issue when the code tries to create the JSON file so it doesn't get created. That in turn gives rise to the "'The credential file does not contain expiration times..." error message since the file does not exist. Here is the exact code I am using (note I put the client_id in an environment variable):
```
# Import the client
from td.client import TDClient
import json
import os

# Create a new session, credentials path is required.
TDSession = TDClient(
    client_id = os.environ['TDAMERITRADE_CLIENT_ID'],
    redirect_uri = 'http://localhost',
    credentials_path='C:/Users/dmcdo/Documents/Trading/td-ameritrade-python-api-0.3.5/samples/credentials.json'
)

# Login to the session
TDSession.login()

# Grab real-time quotes for 'MSFT' (Microsoft)
msft_quotes = TDSession.get_quotes(instruments=['MSFT'])
# Iterate over key/value pairs in dict and print them
for p_id, p_info in msft_quotes.items():
    print("\nTicker:", p_id)
    for key in p_info:
        print(key + ':', p_info[key]) 
```

**Reply by duncanmcdonald** [reply]
I did have trouble originally with the JSON path, but as I mentioned, I debugged it using Thonny which gave me a detailed error message. If you keep having problems I strongly urge you to run your code in Thonny.

**Comment by dougbrowning** [comment]
Edit I tried Thonny and the exact same script worked fine.  I was using PyScrtipter 4.0.  Maybe that is the issue?  Thonny is just so basic, no var type ahead, no var explorer. Unless i am missing something.

I am having the same issue.  No file is created but no error.  Then the message {'credential_path': 'C:\\temp\\td_state.json',
 'message': 'The credential file does not contain expiration times for your '
            'tokens, please go through the oAuth process.'}

I tried all kinds of path locations, used r in front changed the slashes, etc.  

I actually write Python code for a living so I thought I could get it.  

Thanks for any insight.

`# Import the client
from td.client import TDClient

# Create a new session, credentials path is required.
TDSession = TDClient(
    client_id='WMR77777777777777777777777777N3MHDGL9J',
    redirect_uri='http://localhost',
    credentials_path= 'C:/temp/td_state.json',
    auth_flow='flask'
)

# Login to the session
TDSession.login()

# Grab real-time quotes for 'MSFT' (Microsoft)
msft_quotes = TDSession.get_quotes(instruments=['MSFT'])'

**Reply by Cale146** [reply]
@dougbrowning, interesting...  I've been using a text editor in an Ubuntu 20 VM running python 3.8.5 without any luck.  I'm more of a C# guy with limited python experience, but do you have Thonny with python 3.7 as advertised on the official website?  I may have to get the same version as you and see what happens.

**Reply by dougbrowning** [reply]
I downloaded the latest Python 3.9 then installed PyScripter 4.  Standard setup I use at work.  Got nothing.  

I then downloaded Thonny and pointed it to my Python 3.9 I had downloaded before since I had installed the TD api there (from this github).  Ran it and bamm it just worked.  Exact same code as PyScripter.  Kinda of a bummer cause Thonny is super basic.  No idea why but for some reason PyScripter is not creating the file.  Also I am getting a pop up that stops me from being able to copy the URL.

Hope that helps

**Reply by Cale146** [reply]
@dougbrowning , that helps a lot thanks.

**Comment by alwaysftd** [comment]
It's not a permissions issue.  I write python code that writes to these
directories all the time.  It makes no sense and the fact that someone else
is having the same problem finally validates that I'm not crazy.

**Reply by rnneville** [reply]
Did y'all ever solve this issue? I cannot get the td_state.json file created anywhere. It all runs fine, I paste the URL from TDAmeritrade into the console, and then nothing. I'm using pycharm. Help?

**Reply by dougbrowning** [reply]
It all works in Thonny but never can get it to go in PyScripter.  I can even use PyScripter after the file is created by Thonny no prob.  I just had to remake my file since it lasts 90 days and I had to go back to Thonny to get it to go.  It makes no sense but it works.

**Comment by lmsaint** [comment]
I got this to work, but had to use a Jupyter labs notebook(Started server and ran code in browser).  Running it through VS Code never worked.
